src/async_context_frame.cc

```diff
@@ -0,0 +1,260 @@
+  Environment* env = Environment::GetCurrent(info);
+  AsyncContextFrame* acf = Unwrap<AsyncContextFrame>(current(env));
+  info.GetReturnValue().Set(acf->disable(env, info[0]));
+}
+
+void AsyncContextFrame::Exchange(const FunctionCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  info.GetReturnValue().Set(AsyncContextFrame::exchange(env, info[0]));
+}
+
+//
+// Class construction infra
+//
+Local<FunctionTemplate> AsyncContextFrame::GetConstructorTemplate(
+    Environment* env) {
+  return GetConstructorTemplate(env->isolate_data());
+}
+
+Local<FunctionTemplate> AsyncContextFrame::GetConstructorTemplate(
+    IsolateData* isolate_data) {
+  Local<FunctionTemplate> tmpl =
+      isolate_data->async_context_frame_ctor_template();
+  if (tmpl.IsEmpty()) {
+    Isolate* isolate = isolate_data->isolate();
+    tmpl = NewFunctionTemplate(isolate, New);
+    tmpl->InstanceTemplate()->SetInternalFieldCount(
+        BaseObject::kInternalFieldCount);
+    tmpl->SetClassName(FIXED_ONE_BYTE_STRING(isolate, "AsyncContextFrame"));
+    SetProtoMethodNoSideEffect(isolate, tmpl, "get", Get);
+    SetProtoMethodNoSideEffect(isolate, tmpl, "disable", Disable);
+    SetMethod(isolate, tmpl, "current", Current);
+    SetMethod(isolate, tmpl, "disable", DisableStatic);
+    SetMethod(isolate, tmpl, "exchange", Exchange);
+    isolate_data->set_async_context_frame_ctor_template(tmpl);
+  }
+  return tmpl;
+}
+
+bool AsyncContextFrame::HasInstance(Environment* env,
+                                    v8::Local<v8::Value> object) {
+  return GetConstructorTemplate(env->isolate_data())->HasInstance(object);
+}
+
+BaseObjectPtr<AsyncContextFrame> AsyncContextFrame::Create(
+    Environment* env,
+    Local<Value> key,
+    Local<Value> value,
+    Local<Object> current) {
+  Local<Object> obj;
+
+  if (UNLIKELY(!GetConstructorTemplate(env)
+                    ->InstanceTemplate()
+                    ->NewInstance(env->context())
+                    .ToLocal(&obj))) {
+    return BaseObjectPtr<AsyncContextFrame>();
+  }
+
+  return MakeBaseObject<AsyncContextFrame>(env, obj, current, key, value);
+}
+
+void AsyncContextFrame::RegisterExternalReferences(
+    ExternalReferenceRegistry* registry) {
+  registry->Register(New);
+  registry->Register(Get);
+  registry->Register(Disable);
+  registry->Register(DisableStatic);
+  registry->Register(Current);
+  registry->Register(Exchange);
+}
+
+void AsyncContextFrame::CreatePerContextProperties(Local<Object> target,
+                                                   Local<Value> unused,
+                                                   Local<Context> context,
+                                                   void* priv) {
+  Environment* env = Environment::GetCurrent(context);
+
+  auto t = AsyncContextFrame::GetConstructorTemplate(env);
+  SetConstructorFunction(context, target, "AsyncContextFrame", t);
```

Comment:
This line is breaking the cctest `EnvironmentTest.MultipleEnvironmentsPerIsolate` (CI passed because cctest is not built with flag `--with-native-als`).

This can address by declaring a per isolate initialization with `NODE_BINDING_PER_ISOLATE_INIT` and in the list of `NODE_BINDINGS_WITH_PER_ISOLATE_INIT`.